airflow-core/src/airflow/utils/log/log_stream_counter.py

```diff
@@ -0,0 +1,151 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import os
+import tempfile
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from airflow.typing_compat import Self
+    from airflow.utils.log.file_task_handler import (
+        LogHandlerOutputStream,
+        StructuredLogMessage,
+        StructuredLogStream,
+    )
+
+
+class LogStreamCounter:
+    """
+    Memory-efficient log stream counter that tracks the total number of lines while preserving the original stream.
+
+    This class captures logs from a stream and stores them in a buffer, flushing them to disk when the buffer
+    exceeds a specified threshold. This approach optimizes memory usage while handling large log streams.
+
+    Usage:
+
+    .. code-block:: python
+
+        with LogStreamCounter(stream, threshold) as log_counter:
+            # Get total number of lines captured
+            total_lines = log_counter.get_total_lines()
+
+            # Retrieve the original stream of logs
+            for log in log_counter.get_stream():
+                print(log)
+    """
+
+    def __init__(
+        self,
+        stream: LogHandlerOutputStream,
+        threshold: int,
+    ):
+        """
+        Initialize the LogStreamCounter.
+
+        Args:
+            stream: The input log stream to capture and count.
+            threshold: Maximum number of lines to keep in memory before flushing to disk.
+        """
+        self._stream = stream
+        self._threshold = threshold
+        self._buffer: list[StructuredLogMessage] = []
+        self._disk_lines: int = 0
+        self._tmpfile = None
+
+    def _flush_buffer_to_disk(self):
+        """Flush the buffer contents to a temporary file on disk."""
+        if self._tmpfile is None:
+            self._tmpfile = tempfile.NamedTemporaryFile(delete=False, mode="w+", encoding="utf-8")
+
+        self._disk_lines += len(self._buffer)
+        self._tmpfile.writelines(f"{log.model_dump_json()}\n" for log in self._buffer)
+        self._tmpfile.flush()
+        self._buffer.clear()
+
+    def _capture(self):
+        """Capture logs from the stream into the buffer, flushing to disk when threshold is reached."""
+        for log in self._stream:
+            self._buffer.append(log)
+
+            if len(self._buffer) >= self._threshold:
+                self._flush_buffer_to_disk()
```

Comment:
```suggestion
        while True:
            self._buffer = list(islice(self._stream, self._threshold))

            if self._buffer:
                self._flush_buffer_to_disk()
            else:
                break
```

not sure, but might be a little better 🤔

Reply:
Yes, we can use `itertools.isslice` to get up to `self._threshold` lines.
I just refactored with a small change of this function and adjusted the `structured_logs` fixture.